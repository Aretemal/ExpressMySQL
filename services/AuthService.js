import bcrypt from 'bcryptjs';
import jwt from 'jsonwebtoken';
import User from '../models/User.js';
import config from '../config/config.js';

const { secret } = config;

const generationAccessToken = (id) => {
  const payload = { id };
  return jwt.sign(payload, secret, { expiresIn: '24h' });
};
class AuthService {
  async registration({
    userName: login, password, firstName, lastName, email,
  }) {
    const candidate = await User.findOne({
      where: {
        login,
      },
    });
    if (candidate) {
      throw new Error('Пользователь уже существует');
    }
    const hashPassword = bcrypt.hashSync(password, 7);
    await User.create({
      login, password: hashPassword, firstName, lastName, email,
    });
    return { message: 'Пользователь зарегистрирован' };
  }

  async login({ login, password }) {
    const user = await User.findOne({ where: { login } });
    if (!user) {
      throw new Error(`Пользователь ${login} не найден`);
    }
    const validPassword = bcrypt.compareSync(password, user.password);
    if (!validPassword) {
      throw new Error('Введен неверный пароль');
    }
    const token = generationAccessToken(user.id);
    return token;
  }
}

export default new AuthService();
