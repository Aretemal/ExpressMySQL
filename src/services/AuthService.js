import bcrypt from 'bcryptjs';
import jwt from 'jsonwebtoken';
import * as dotenv from 'dotenv';
import User from '../models/user.js';

dotenv.config();

const generationAccessToken = (id) => {
  const payload = { id };
  return jwt.sign(payload, process.env.SECRET_KEY_RANDOM, { expiresIn: '24h' });
};
class AuthService {
  async registration(
    {
      login, password, firstName, lastName, email,
    },
    next,
  ) {
    const candidate = await User.findOne({
      where: {
        login,
      },
    });
    if (candidate) {
      next({ errorsArray: [{ msg: 'User already exists' }] });
      return;
    }
    const hashPassword = bcrypt.hashSync(password, 7);
    const user = await User.create({
      login, password: hashPassword, firstName, lastName, email,
    });
    const token = generationAccessToken(user.id);
    return { token, id: user.id };
  }

  async login({ login, password }, next) {
    const user = await User.findOne({ where: { login } });
    if (!user) {
      next({ errorsArray: [{ msg: 'User not found' }] });
      return;
    }
    const validPassword = bcrypt.compareSync(password, user.password);
    if (!validPassword) {
      next({ errorsArray: [{ msg: 'Wrong password entered' }] });
      return;
    }
    const token = generationAccessToken(user.id);
    return { token, id: user.id };
  }
}

export default new AuthService();
