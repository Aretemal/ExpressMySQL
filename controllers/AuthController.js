import { validationResult } from 'express-validator';
import AuthService from '../services/AuthService.js';

class AuthController {
  async registration(req, res) {
    try {
      const errors = validationResult(req);
      if (!errors.isEmpty()) {
        return res.status(400).json({ message: 'Ошибка регистрациии' });
      }
      const result = await AuthService.registration(req.body);
      return result;
    } catch (e) {
      console.log(e);
      res.status(400).json({ message: 'Registration error' });
    }
  }

  async login(req, res) {
    try {
      const token = await AuthService.login(req.body);
      return res.json({ token });
    } catch (e) {
      console.log(e);
      res.status(400).json({ message: 'Login error' });
    }
  }
}

export default new AuthController();
