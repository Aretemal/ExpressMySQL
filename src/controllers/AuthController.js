import { validationResult } from 'express-validator';
import UserSerializer from '../serializers/UserSerializer.js';
import AuthService from '../services/AuthService.js';
import AuthSerializer from '../serializers/AuthSerializer.js';

class AuthController {
  async registration(req, res, next) {
    const errors = validationResult(req);
    if (!errors.isEmpty()) {
      return res.status(400).json({ errors: errors.array() });
    }
    const user = await AuthService.registration(req.body);
    req.serializer = new UserSerializer(user);
    next();
  }

  async login(req, res, next) {
    const errors = validationResult(req);
    if (!errors.isEmpty()) {
      return res.status(400).json({ errors: errors.array() });
    }
    const token = await AuthService.login(req.body);
    req.serializer = new AuthSerializer(token);
    next();
  }
}

export default new AuthController();
