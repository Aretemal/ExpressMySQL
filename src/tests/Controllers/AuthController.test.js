/* eslint-disable no-undef */
// NODE_OPTIONS=--experimental-vm-modules npx jest
import sinon from 'sinon';
import AuthController from '../../controllers/AuthController';
import AuthService from '../../services/AuthService.js';
import { req, res } from './ReqRes.js';

describe('Auth Controller : ', () => {
  test('Creates a new user', async () => {
    const stub = sinon.stub(AuthService, 'registration');
    stub.returns({
      userId: 1, login: 'Artem', firstName: 'Artem', lastName: 'Novik',
    });
    await AuthController.registration(req, res);
    expect(res.dataJS.data.attributes.attributes.userId).toBe(1);
    expect(res.dataJS.data.attributes.attributes.firstName).toBe('Artem');
    expect(res.dataJS.data.attributes.attributes.lastName).toBe('Novik');
  });
  test('Authorization', async () => {
    const stub = sinon.stub(AuthService, 'login');
    stub.returns({
      user: { id: 1 }, token: 'token',
    });
    await AuthController.login(req, res);
    expect(res.dataJS.data.attributes.attributes.token).toBe('token');
    expect(res.dataJS.data.attributes.attributes.user).toBeDefined();
  });
});
