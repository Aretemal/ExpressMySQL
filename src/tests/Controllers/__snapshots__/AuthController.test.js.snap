// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Auth Controller : Login should return token and id 1`] = `
AuthSerializer {
  "resource": {
    "id": 1,
    "token": "token",
  },
}
`;

exports[`Auth Controller : Registration should create a new user and return token and id 1`] = `
AuthSerializer {
  "resource": {
    "id": 1,
    "token": "token",
  },
}
`;
